File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-1.swift
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     var val: Int
 *     var next: ListNode?
 *     init(_ val: Int) {
 *         self.val = val
 *         self.next = nil
 *     }
 * }
 */

class Solution {
    func mergeTwoLists(_ list1: ListNode?, _ list2: ListNode?) -> ListNode? {
        var dummyHead = ListNode(-1)
        var curr = dummyHead

        var list1 = list1
        var list2 = list2

        while let l1 = list1, let l2 = list2 {
            if l1.val < l2.val {
                curr.next = l1
                list1 = l1.next
                curr = l1
            } else {
                curr.next = l2
                list2 = l2.next
                curr = l2
            }
        }

        if let l1 = list1 {
            curr.next = l1
        } else if let l2 = list2 {
            curr.next = l2
        }

        return dummyHead.next
    }
}

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def merge2Lists(self, h1: Optional[ListNode], h2: Optional[ListNode]) -> Optional[ListNode]:
        dummyHead = ListNode()
        curr = dummyHead

        while h1 and h2:
            if h1.val < h2.val:
                curr.next = h1
                h1 = h1.next
            else:
                curr.next = h2
                h2 = h2.next
            curr = curr.next

        if h1:
            curr.next = h1
        elif h2:
            curr.next = h2
        
        return dummyHead.next

    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        if len(lists) < 1:
            return None

        while len(lists) > 1:
            mergedLists = []

            for i in range(0, len(lists), 2):
                l1 = lists[i]
                l2 = lists[i + 1] if i+1 <len(lists) else None
                mergedLists.append(self.merge2Lists(l1, l2))
            lists = mergedLists
        return lists[0]

        